sound/soc/sof/intel/hda-dai.c

```diff
@@ -122,6 +145,64 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
 	return 0;
 }
 
+static int hda_link_config_ipc(struct sof_intel_hda_stream *hda_stream,
+			       const char *dai_name, int channel, int dir)
+{
+	struct sof_ipc_dai_config *config;
+	struct snd_sof_dai *sof_dai;
+	struct sof_ipc_reply reply;
+	char *sof_dai_name;
+	int ret = 0;
+
+	list_for_each_entry(sof_dai, &hda_stream->sdev->dai_list, list) {
+		if (!sof_dai->name)
+			continue;
+
+		/*
+		 * TODO: this looks a bit hacky. Need to save cpu dai name in
+		 * snd_sof_dai to avoid this
+		 */
+		sof_dai_name = kasprintf(GFP_KERNEL, "%s Pin", sof_dai->name);
+		if (!sof_dai->name)
+			return -EINVAL;
+
+		if (!strcmp(dai_name, sof_dai_name)) {
+			if (dir == sof_dai->comp_dai.direction) {
+				config = sof_dai->dai_config;
+
+				if (!config) {
+					dev_err(hda_stream->sdev->dev,
+						"error: no config for DAI %s\n",
+						sof_dai->name);
+					return -EINVAL;
+				}
+
+				/* update config with stream tag */
+				config->hda.link_dma_ch = channel;
+
+				/* send IPC */
+				ret = sof_ipc_tx_message(hda_stream->sdev->ipc,
+							 config->hdr.cmd,
+							 config,
+							 config->hdr.size,
+							 &reply, sizeof(reply));
+
+				if (ret < 0)
+					dev_err(hda_stream->sdev->dev,
+						"error: failed to set dai config for %s\n",
+						sof_dai->name);
+
+				kfree(sof_dai_name);
+				return ret;
+			}
+		}
+
+		kfree(sof_dai_name);
+	}
+
+	return ret;
```

Comment:
this is just massively complicated and hard to review.

Reply:
@plbossart no doubt this is complicated. It took me a while to figure out how to make it work for the N:M case and then implement it while running into some vicious issues with string manipulation. If implementing it was hard, I can only imagine the review to be even harder. 

I probably need to make a design document to explain the methodology to make it easier to review. Let me work on it today.